src/librustc/middle/lang_items.rs

```diff
@@ -298,68 +322,68 @@ language_item_table! {
     // defined to use it, but a final product is required to define it
     // somewhere. Additionally, there are restrictions on crates that use a weak
     // lang item, but do not have it defined.
-    PanicFnLangItem,                 "panic",                   panic_fn;
-    PanicBoundsCheckFnLangItem,      "panic_bounds_check",      panic_bounds_check_fn;
-    PanicInfoLangItem,               "panic_info",              panic_info;
-    PanicImplLangItem,               "panic_impl",              panic_impl;
+    PanicFnLangItem,             "panic",              panic_fn,                Target::Fn;
+    PanicBoundsCheckFnLangItem,  "panic_bounds_check", panic_bounds_check_fn,   Target::Fn;
+    PanicInfoLangItem,           "panic_info",         panic_info,              Target::Struct;
```

Comment:
In the future there might be a language item that could possibly be some sort of type definition, not specifically a `struct`.

This is especially notable here. `panic_info` in `no_core` could easily be defined to be an `enum` as well for example – in fact, the only reason this is a language item is to facilitate type-checking the `panic_impl` function.

Yet, I do not see a nice way out here, and since language items are not stable anyway it seems fine to keep it restrictive for now.

Reply:
I think this could be extended to have a `Target::Any` that allows the attribute to be applied to anything, as was the behaviour before this PR. 

Alternatively, the definition of `Target` could be changed to a `u32` and have each bit represent a variant (similar to `AdtFlags`) and then the language item definitions could `|` together these “variants” to indicate that the attribute could be applied to either construct. 

I agree that, for the moment, the simplest solution would be to keep this more restrictive approach and change if necessary.